x/incentive/types/tx.go

```diff
@@ -0,0 +1,317 @@
+		Asset:    asset,
+	}
+}
+
+func (msg MsgBeginUnbonding) Route() string { return sdk.MsgTypeURL(&msg) }
+func (msg MsgBeginUnbonding) Type() string  { return sdk.MsgTypeURL(&msg) }
+
+func (msg *MsgBeginUnbonding) ValidateBasic() error {
+	return validateSenderAssetTier(msg.Supplier, msg.Tier, &msg.Asset)
+}
+
+func (msg *MsgBeginUnbonding) GetSigners() []sdk.AccAddress {
+	return checkers.Signers(msg.Supplier)
+}
+
+// GetSignBytes get the bytes for the message signer to sign on
+func (msg *MsgBeginUnbonding) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func NewMsgSponsor(sponsor sdk.AccAddress, programID uint64, asset sdk.Coin) *MsgSponsor {
+	return &MsgSponsor{
+		Sponsor: sponsor.String(),
+		Program: programID,
+		Asset:   asset,
+	}
+}
+
+func (msg MsgSponsor) Route() string { return sdk.MsgTypeURL(&msg) }
+func (msg MsgSponsor) Type() string  { return sdk.MsgTypeURL(&msg) }
+
+func (msg *MsgSponsor) ValidateBasic() error {
+	if msg.Program == 0 {
+		return ErrInvalidProgramID.Wrapf("%d", msg.Program)
+	}
+	return validateSenderAsset(msg.Sponsor, &msg.Asset)
+}
+
+func (msg *MsgSponsor) GetSigners() []sdk.AccAddress {
+	return checkers.Signers(msg.Sponsor)
+}
+
+// GetSignBytes get the bytes for the message signer to sign on
+func (msg *MsgSponsor) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func NewMsgCreateProgram(authority, title, description string, program IncentiveProgram) *MsgCreateProgram {
+	return &MsgCreateProgram{
+		Title:       title,
+		Description: description,
+		Program:     program,
+		Authority:   authority,
+	}
+}
+
+func (msg MsgCreateProgram) Route() string { return sdk.MsgTypeURL(&msg) }
+func (msg MsgCreateProgram) Type() string  { return sdk.MsgTypeURL(&msg) }
+
+func (msg MsgCreateProgram) String() string {
+	out, _ := yaml.Marshal(msg)
+	return string(out)
+}
+
+func (msg *MsgCreateProgram) GetTitle() string { return msg.Title }
+
+func (msg *MsgCreateProgram) GetDescription() string { return msg.Description }
+
+func (msg MsgCreateProgram) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(msg.Authority); err != nil {
+		return sdkerrors.Wrap(err, "invalid authority address")
+	}
+
+	if err := msg.ValidateAbstract(); err != nil {
+		return err
+	}
+
+	/*
+		TODO: General validation
+		if err := msg.Program.Validate(); err != nil {
+			return err
+		}
+	*/
+
+	// Additional rules apply to incentive programs which are still being proposed
+	if msg.Program.Id != 0 {
+		return ErrInvalidProgramID.Wrapf("%d", msg.Program.Id)
+	}
+	if !msg.Program.RemainingRewards.IsZero() {
+		return ErrNonzeroRemainingRewards.Wrap(msg.Program.RemainingRewards.String())
+	}
+	if !msg.Program.FundedRewards.IsZero() {
+		return ErrNonzeroFundedRewards.Wrap(msg.Program.FundedRewards.String())
+	}
+
+	return nil
+}
+
+func (msg *MsgCreateProgram) ValidateAbstract() error {
+	title := msg.GetTitle()
```

Comment:
let's remove getters (see example in x/leverage proto files) to make the code smaller.

Reply:
done